src/approximate-qvm.lisp

```diff
@@ -0,0 +1,124 @@
+
+
+(defun (setf qubit-t2) (t2 qvm qubit)
+  "Save the kraus operators for the t2 provided on the specified qubit."
+  (let ((kraus-ops (generate-damping-dephasing-kraus-ops t2 (avg-gate-time qvm))))
+    (check-kraus-ops kraus-ops)
+    (setf (gethash qubit (t2-ops qvm)) kraus-ops))))
+
+
+(defmethod (setf qubit-fro) (fro qvm qubit)
+  "Save the readout povms for the fro provided on the specified qubit. The fR0 is assumed to be the average readout fidelity (avg of p(1|1) and p(0|0)) so as an approximation, we set p(0|0) = p(1|1) = fR0 in the readout povm assignment probabilities. "
+  (setf (gethash qubit (readout-povms qvm)) (list fro (- 1.0d0 fro) (- 1.0d0 fro) fro)))
+
+
+; For now just use a single gate time value for all calculations
+(defmethod set-gate-time ((qvm approx-qvm) gate-time)
+  "Set the gate time slot for approx-qvm. This value should represent the average gate time of the gates that will be run."
+  (setf (avg-gate-time qvm) gate-time))
+
+
+(defmethod apply-all-kraus-operators ((qvm approx-qvm) (instr quil:gate-application) qubits)
+  "Apply all the kraus operators (for all of the noise sources) to the state of the program. If we have just encountered a 2-qubit gate, we need to tensor the kraus operators for each qubit's error source. Currently, this does not support > 2 qubit gates. Note: if the current instr is a 2 qubit gate and there is only kraus ops for one of the qubits, then no error is applied. For example, if q0 has T1 error associated with it but q1 does not, no error channel is applied after a CNOT 0 1. "
+  (loop :for source :in (list (t1-ops qvm) (t2-ops qvm))
+        :do (cond ((= (length qubits) 1) 
+                   (let ((kraus-map (gethash (nth 0 qubits) source)))
+                     (when kraus-map (apply-kraus-ops qvm instr kraus-map) )))
+                  ((= (length qubits) 2) 
+                   (let ((kraus-map  (kraus-kron (gethash (nth 0 qubits) source) 
+                                                 (gethash (nth 1 qubits) source))))
+                     (when kraus-map (check-kraus-ops kraus-map) (apply-kraus-ops qvm instr kraus-map))))              
+                  ((> (length qubits) 2) 
+                   (error "Unable to APPLY-ALL-KRAUS-OPERATORS: expected two or less qubits, but got ~A" (length qubits))))))
+
+
+(defmethod transition ((qvm approx-qvm) (instr quil:gate-application))
+  "For the current instruction in the program, apply the gate corresponding to that
+instruction, and then apply all the kraus operators for all the sources of noise following
+that instruction. "
+  (let ((gate   (pull-teeth-to-get-a-gate instr))
+        (params (mapcar (lambda (p) (force-parameter p qvm))
+                        (quil:application-parameters instr)))
+        (qubits (mapcar #'quil:qubit-index (quil:application-arguments instr))))
+    (apply #'apply-gate gate (amplitudes qvm) (apply #'nat-tuple qubits) params)
+    (apply-all-kraus-operators qvm instr qubits)
```

Comment:
The generic method for transition is just a pass-through. Not sure how to answer your last question :(. I thought that specifying instr is a `quil:gate-application` would be the right thing to do because I only want to apply gates (and potentially kraus operators) after gate application instructions.

Reply:
This is a fine answer: I was worried that there _wasn't_ a fully generic definition for this function, in which case passing a MEASURE would throw an error. If there is a generic definition, I feel all set.